botocore/credentials.py

```diff
@@ -791,6 +821,157 @@ def _assume_role_base_kwargs(self, config):
+                      {"username": "username", "password": "password"})
+            assertion = SamlFormsBasedAuthenticator().authenticate(
+                config['saml_endpoint'],
+                {mapping["username"]: config['saml_username'],
+                 mapping["password"]: self.password_prompter("Password: ")},
+                verify=config.get('saml_verify_ssl') != 'false')
+        else:
+            raise ValueError("Unsupported saml_authentication_type: %s"
+                             % config.get('saml_authentication_type'))
+        if not assertion:
+            raise ValueError('Login failed: SAML assertion not found')
+        idp_roles = self._parse_roles(assertion)
+        if not idp_roles:
+            raise ValueError('Identity provider provides no role.')
+        role = self.role_selector(config.get('role_arn'), idp_roles)
+        if not role:
+            raise ValueError('Unable to choose role "%s" from %s' % (
+                config.get('role_arn'), [r['RoleArn'] for r in idp_roles]))
+        role['SAMLAssertion'] = assertion
+        return role
+
+    def _parse_roles(self, assertion):
+        attribute = '{urn:oasis:names:tc:SAML:2.0:assertion}Attribute'
+        attr_value = '{urn:oasis:names:tc:SAML:2.0:assertion}AttributeValue'
+        awsroles = []
+        root = ET.fromstring(base64.b64decode(assertion))
+        for attr in root.getiterator(attribute):
+            if attr.get('Name')=='https://aws.amazon.com/SAML/Attributes/Role':
+                for value in attr.getiterator(attr_value):
+                    parts = value.text.split(',')
+                    # Deals with "role_arn,pricipal_arn" or its reversed order
+                    if 'saml-provider' in parts[0]:
+                        role = {'PrincipalArn': parts[0], 'RoleArn': parts[1]}
+                    else:
+                        role = {'PrincipalArn': parts[1], 'RoleArn': parts[0]}
+                    awsroles.append(role)
+        return awsroles
+
+
+class SamlFormsBasedAuthenticator(object):
+    def authenticate(self, endpoint, params, verify=True):
+        login_form = self._get_form(requests.get(endpoint, verify=verify).text)
+        if login_form is None:
+            raise ValueError('Login form is not found in %s' % endpoint)
+        payload = dict((tag.attrib['name'], tag.attrib.get('value', ''))
+                       for tag in login_form.findall(".//input"))
+        for field, value in params.items():
+            if field in payload:
+                payload[field] = value
+        response_form = self._get_form(requests.post(
+            urljoin(endpoint, login_form.attrib.get('action', '')),
+            data=payload, verify=verify).text)
+        if response_form is not None:
+            return self._get_value_of_first_tag(
+                response_form, 'input', 'name', 'SAMLResponse')
+        # Login failed, typically caused by incorrect username and/or password
+        return None
```

Comment:
It would be hard to debug the reason and present it to user, because the login fail page is totally undefined in SAML standard. That being said, for a forms-based authentication, user can always open the endpoint url in browser and manually login and see what goes wrong. I'll change that "assertion not found" exception message to include the endpoint url, so that the user will be encouraged to try it out by themselves.